src/fabric/executor.rs

```diff
@@ -137,32 +137,26 @@ impl Executor {
         assert!(prev.is_none(), "duplicate result id: {id:?}");
 
         // Execute any ready dependencies
-        for op_id in self
-            .dependencies
-            .get(id)
-            .cloned()
-            .unwrap_or_default()
-            .into_iter()
-        {
-            {
-                let mut operation = self.operations.get_mut(op_id).unwrap();
+        if let Some(deps) = self.dependencies.get(id) {
+            let mut ready_ops = Vec::new();
+            for op_id in deps.iter() {
+                {
+                    let mut operation = self.operations.get_mut(*op_id).unwrap();
+
+                    operation.inflight_args -= 1;
+                    if operation.inflight_args > 0 {
+                        continue;
+                    }
+                } // explicitly drop the mutable `self` reference
 
-                operation.inflight_args -= 1;
-                if operation.inflight_args > 0 {
-                    continue;
-                }
-            } // explicitly drop the mutable `self` reference
-
-            // Take ownership of the operation
-            let op = self.operations.take(op_id).unwrap();
-
-            // Get the inputs and execute the method to produce the output
-            let inputs = op
-                .args
-                .iter()
-                .map(|id| self.results.get(*id).unwrap().value.clone())
-                .collect::<Vec<_>>();
-            self.execute_operation(op, inputs);
+                // Mark the operation as ready for execution
+                ready_ops.push(*op_id);
+            }
+
+            for op in ready_ops.into_iter() {
+                let op = self.operations.take(op).unwrap();
+                self.execute_operation(op);
+            }
```

Comment:
i must be missing something here, but it seems like we just moved the troublesome `.map(|id| self.results.get(*id).unwrap().value.clone())` line into `execute_operation`, but still call `execute_operation` the same number of times (since we push into `read_ops` on every iteration of the `deps.iter()` loop). does this actually reduce the amount of clones?